src/Domain/Download/FolderTreeDownloader.php

```diff
@@ -65,37 +65,52 @@ private function createZip(Dossier $folder): void
-        );
-
-        foreach ($subFolders as $folder) {
-            $this->addFolderContentRecursively($zip, $folder, sprintf('%s/%s', $folderPath, $this->sanitizeNode($folder->getNom())));
+        $folders = $this->getFoldersInFolder($beneficiary, $folder);
+        foreach ($folders as $folder) {
+            $this->addFolderContentRecursively($zip, $beneficiary, $folder, sprintf('%s/%s', $folderPath, $this->sanitizeNode($folder->getNom())));
         }
     }
 
-    public function addFolderTreeRecursively(ZipStream $zip, Beneficiaire $beneficiary, string $folderPath = ''): void
+    /**
+     * @return Collection<int, Document>
+     */
+    private function getDocumentsInFolder(Beneficiaire $beneficiary, ?Dossier $folder = null): Collection
     {
-        $documents = $beneficiary->getDocuments()->filter(
-            fn (Document $document) => is_resource($this->bucketService->getObjectStream($document->getObjectKey())) && !$document->getDossier()
-                && $this->authorizationChecker->isGranted(PersonalDataVoter::DOWNLOAD, $document)
-        );
-        $this->addDocuments($zip, $documents, $folderPath);
-
-        $subFolders = $beneficiary->getDossiers()->filter(
-            fn (Dossier $folder) => !$folder->hasParentFolder() && $this->authorizationChecker->isGranted(PersonalDataVoter::DOWNLOAD, $folder),
-        );
+        return $folder
+            ? $folder
+                ->getDocuments($this->authorizationChecker->isGranted(PersonalDataVoter::DOWNLOAD, $folder))
+                ->filter(
+                    fn (Document $document) => is_resource($this->bucketService->getObjectStream($document->getObjectKey()))
+                )
+            : $beneficiary
+                ->getDocuments()
+                ->filter(
+                    fn (Document $document) => is_resource($this->bucketService->getObjectStream($document->getObjectKey())) && !$document->getDossier()
+                        && $this->authorizationChecker->isGranted(PersonalDataVoter::DOWNLOAD, $document)
+                );
+    }
 
-        foreach ($subFolders as $folder) {
-            $this->addFolderContentRecursively($zip, $folder, sprintf('%s/%s', $folderPath, $this->sanitizeNode($folder->getNom())));
-        }
+    /**
+     * @return Collection<int, Dossier>
+     */
+    private function getFoldersInFolder(Beneficiaire $beneficiary, ?Dossier $folder = null): Collection
+    {
+        return $folder
+            ? $folder
+                ->getSousDossiers()
+                ->filter(
+                    fn (Dossier $folder) => $this->authorizationChecker->isGranted(PersonalDataVoter::DOWNLOAD, $folder)
+                )
+            : $beneficiary
+                ->getDossiers()
+                ->filter(
+                    fn (Dossier $folder) => !$folder->hasParentFolder() && $this->authorizationChecker->isGranted(PersonalDataVoter::DOWNLOAD, $folder),
+                );
```

Comment:
I'm not sure it's better but we could write it like this : 
```
$folders = $folder ? $folder->getSousDossiers() : $beneficiary->getRootFolders();
return $folders->filter(fn (Dossier $folder) => $this->authorizationChecker->isGranted(PersonalDataVoter::DOWNLOAD, $folder));
```

Reply:
@NilsRec It's clearer this way i think too. This code `!$folder->hasParentFolder()` is it in the getRootFolders() method in your case ?